### ISSUE TITLE
scala classes as parameter types

### ISSUE THREAD

**Issue by chandu0101** [payalabs/scalajs-react-bridge#4]
**Case1**

 i have MuiMenu component ,it accepts a array of js object  for menuItems prop,

its more safe to accept  ``menuItems : JArray[MuiMenuItem]`` instead of ``JArray[js.Object]``

and in macro we just do 
   `` p.updateDynamic("menuItems")(menuItems.map(_.toJson))``

it'll be simple to implement , one way is we just define a trait JSScalaType with an abstract method toJson ... and end user must extend that .. (Note: names are arbitary .. )

**Case2**

some props of js component accepts only few values  of a type 

```
menuType : ReactProps.oneOf(LINK,SUBHEADER)

```

in my code i am using value classes for this purpose ..

```scala

menuType: js.UndefOr[MuiMenuItemType] = js.undefined


class MuiMenuItemType private(val name: String) extends AnyVal

object MuiMenuItemType {

  val SUBHEADER = new MuiMenuItemType("SUBHEADER")
  val LINK = new MuiMenuItemType("LINK")
  val NESTED = new MuiMenuItemType("NESTED")

  def newType(name: String) = new MuiMenuItemType(name)


}

// while passing to js comp i am accessing its public field
    menuType.foreach(v => p.updateDynamic("menuType")(v.name))

```
any thoughts on how we can handle this case2 ..

**Comment by ramnivas** [comment]
I believe both may be addressed with the same technique (essentially what you suggested + a macro). Here is outline to see if this will work for your use cases.

- Define an abstract class `ReactBridgeObject` with a `toJs: js.Any` method implemented using macro to return an object literal (the same way I calculate props in `ReactBridgeComponent`).
- Users can extend this class if its structure maps directly to the expected dynamic object's schema. Perhaps, that is all needed for case 1: `class MuiMenuItem(name: String) extends ReactBridgeObject`
- Users can extends `ReactBridgeObject`, but override `toJs` to whatever is appropriate if the default implementation doesn't suite the use case. I believe this will take care of case 2.

**Comment by chandu0101** [comment]
> I believe this will take care of case 2.

yeah same approach works , but i have a doubt about creating instances for valuetypes if i use base trait...

**Comment by ramnivas** [comment]
Fixed via 67f80b0a623ec8a3c5a550e174c2a2b0418b92e7. Thanks @chandu0101

**Comment by chandu0101** [comment]
its funny that i missed my use case mentioned in this post , what if i pass array[objects] i didn't implemented this . 
```scala
 private def toArray: PartialFunction[Any, js.Array[_]] = {
    case oSeq: Seq[ReactBridgeObject] => js.Array(oSeq: _*).map(_.toJs)
    case seq: Seq[_] => js.Array(seq: _*)
  }
```

i believe this should work , but i am getting compiler warning 

`` non-variable type argument com.payalabs.scalajs.react.bridge.ReactBridgeObject in type pattern Seq[com.payalabs.scalajs.react.bridge.ReactBridgeObject] (the underlying of Seq[com.payalabs.scalajs.react.bridge.ReactBridgeObject]) is unchecked since it is eliminated by erasure
``
do i have to create new converter which will accept exact type or is there any other way .. ?

**Comment by ramnivas** [comment]
I believe I will have to raise the priority for #2. By the time we reach runtime, erasure has removed too much info for us to do anything too interesting. I hope to get some time soon to deal with it.

Meanwhile, if this is a show-stopper for you, can you change the parameter to something that extend `ReactBridgeObject` (perhaps along with an implicit conversion from `Seq[<your-type>]`)?

**Comment by chandu0101** [comment]
nah its not a stopper for me ,take u r own time :+1:

**Comment by ramnivas** [comment]
@chandu0101 Your use case is now fixed in a more general manner (see 3be28c92ab3ba1fb8256276d93a06c171c96cbfa). See tests to see how it can be used.

**Comment by chandu0101** [comment]
nice! ,its more cleaner now :+1:
